go/plugins/vertexai/vertexai.go

```diff
@@ -213,13 +211,77 @@ func (g *generator) generate(ctx context.Context, input *ai.GenerateRequest, cb
 	// TODO: gm.ToolConfig?
 
 	// Send out the actual request.
-	resp, err := cs.SendMessage(ctx, parts...)
-	if err != nil {
-		return nil, err
+	if cb == nil {
+		resp, err := cs.SendMessage(ctx, parts...)
+		if err != nil {
+			return nil, err
+		}
+
+		r := translateResponse(resp)
+		r.Request = input
+		return r, nil
 	}
 
-	r := translateResponse(resp)
-	r.Request = input
+	// Streaming version.
+	iter := cs.SendMessageStream(ctx, parts...)
+	r := &ai.GenerateResponse{Request: input, Candidates: make([]*ai.Candidate, input.Candidates)}
+	for {
+		chunk, err := iter.Next()
+		if err != nil {
+			if err.Error() == "no more items in iterator" {
+				break
+			}
+			return nil, err
+		}
+
+		// Process each candidate.
+		for _, c := range chunk.Candidates {
+			tc := translateCandidate(c)
+
+			// Call callback with the candidate info.
+			err := cb(ctx, &ai.GenerateResponseChunk{
+				Content: tc.Message.Content,
+				Index:   tc.Index,
+			})
+			if err != nil {
+				return nil, err
+			}
+			// Save candidate in full response structure.
+			if old := r.Candidates[tc.Index]; old == nil {
+				r.Candidates[tc.Index] = tc
+			} else {
+				// Need to merge two "parts" of a candidate.
```

Comment:
Would it help if vertexai/genai.joinResponses was exported?

Reply:
Yes, probably would. I think the general state of streaming documentation is pretty poor, so it's hard to know for sure.

Reply:
That's a good reference though, I can just copy what they do.

Reply:
They just need to export the merged field somehow, they are already calling it. I think you did the same for googleai?
It's almost accessible, I can get the content field from the chat history. But the rest is inaccessible.

Reply:
See https://github.com/googleapis/google-cloud-go/pull/10364

Reply:
Using the newly exported api.